### PR TITLE
Implement named threads on Windows

### PR DESCRIPTION
**Pull request by jsheard**
Adds support for named threads on Windows, using the very awkward convention supported by most debuggers.

example

### CODE REVIEW

**Comment by jsheard** [comment]
Hang on, this is broken on 32-bit. I'll fix it tomorrow.

**Comment by retep998** [comment]
The stack overflow handler is not always guaranteed to be registered, for example cdylibs that are loaded and invoked from other languages. It would perhaps be better to have a separate handler for this that is lazily registered when you set the name.

**Comment by jsheard** [comment]
Good point @retep998, I've moved it to a separate handler.

32-bit is still broken, I think the `THREADNAME_INFO` struct isn't being packed correctly. The official definition declares it with `#pragma pack(push,8)` but as far as I can tell there's no equivalent modifier in Rust. Sorry if I'm missing something obvious.

**Comment by alexcrichton** [comment]
Perhaps the 32-bit definition could manually fill in padding/packing/etc?

**Comment by retep998** [comment]
I just tested the struct from the example and as far as I can tell the `#pragma pack(push,8)` has no effect on the size or alignment or layout of the struct, neither on 32bit nor 64bit.

File: src/libstd/sys/windows/thread.rs [hunk 1]
```diff
@@ -53,11 +68,36 @@ impl Thread {
...
+            szName: name.as_ptr(),
+            dwThreadID: 0xffffffff,
+            dwFlags: 0,
+        };
+
+        let size = (mem::size_of::<c::THREADNAME_INFO>() /
+                    mem::size_of::<c::ULONG_PTR>()) as c::DWORD;
+
+        unsafe {
+            c::RaiseException(c::MS_VC_EXCEPTION, 0, size, mem::transmute(&info));
```
Comment (retep998) [inline src/libstd/sys/windows/thread.rs]: Please don't use transmute. You can safely cast raw pointers `&info as *const _ as *const _`.

**Comment by jsheard** [comment]
32-bit is now working 🎉

The struct packing was actually fine, the problem was caused by a pre-existing bug. `ULONG_PTR` is supposed to be a pointer-sized unsigned integer but `c.rs` currently defines it as an alias for `c_ulonglong/u64`, so `set_name` was getting an invalid result from `size_of::<ULONG_PTR>()` in 32-bit builds.

**Comment by alexcrichton** [comment]
Whoa nice catch @jsheard! It looks like there's a few other locations in the codebase that define `ULONG_PTR`, mind updating those as well?

**Comment by alexcrichton** [comment]
Awesome thanks @jsheard and thanks with the implementation! Want to squash commits and I'll r+?

**Comment by jsheard** [comment]
I just realized there's a serious bug in my original approach. The magic 0x406D1388 exception code is **not** reserved for passing messages to the debugger, so if some linked C/C++ code happens to use 0x406D1388 for some other purpose the global exception filter would silently eat their exceptions 😱

I've pushed a different implementation that should be more robust. Does this look sensible?

**Comment by retep998** [comment]
If only Rust exposed a way to catch SEH exceptions the way you can in Microsoft C...

**Comment by jsheard** [comment]
We don't strictly _need_ to check the exception code, since `IGNORE_EXCEPTIONS` is only ever `true` when the exception code is `MS_VC_EXCEPTION`. That said it might be more efficient to check the code anyway and early-return if it isn't `MS_VC_EXCEPTION` to avoid touching the thread-local storage unnecessarily.

**Comment by alexcrichton** [comment]
From the [documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms681419(v=vs.85).aspx) vectored handlers:

> The handler should not call functions that acquire synchronization objects or allocate memory, because this can cause problems. Typically, the handler will simply access the exception record and return.

Accessing TLS, however, will on Windows possibly allocate memory and acquire synchronization objects, so using TLS here may not play out so well :(

**Comment by jsheard** [comment]
Argh, I'm out of ideas.

Perhaps we should just abandon the magic exception and focus on [SetThreadDescription.](https://msdn.microsoft.com/en-us/library/windows/desktop/mt774976(v=vs.85).aspx) It has the downside of only working on Windows 10 and literally no tools actually support it yet (not even Visual Studio), but time will solve that problem eventually.

**Comment by alexcrichton** [comment]
I'd personally be ok with the original implementation, it seems like another excepting colliding with this one is exceedingly rare because it'd break anyone attempting to set the name of a thread like this, right?

**Comment by jsheard** [comment]
@alexcrichton most programs wouldn't set a thread name like this, in C or C++ you would use Microsoft's SEH language extension to catch the exception in a scoped block like this:

```
__try {
    RaiseException(MS_VC_EXCEPTION, 0, sizeof(info) / sizeof(ULONG_PTR), (ULONG_PTR*)&info);  
} __except (EXCEPTION_CONTINUE_EXECUTION) {}
```

This is easy and safe, but impossible to express in pure Rust so we're in the awkward position of trying to use program-wide VEH callbacks without causing any side effects in the rest of the program.

**Comment by alexcrichton** [comment]
If you're feeling intrepid you could actually implement that in Rust! I could help you out poking around in the compiler if you'd like. This'd basically just surface itself as a new intrinsic that's MSVC specific and also specific to the standard library. I'm not 100% sure how much work it would take, though. I believe LLVM is capable of producing the code to match the semantics of that block, so we'd just need to harness it.

**Comment by jsheard** [comment]
I'm sure it's possible, but it would be an insane amount of work and complexity just to implement this one trivial feature. IMO there's no practical and 100% safe way to implement this via the exception.

I'll return to this when debuggers start using the new `Set/GetThreadDescription` APIs.

**Comment by alexcrichton** [comment]
Ok that's also fine by me, thanks regardless for the PR @jsheard!

**Comment by jsheard** [comment]
Already did! It's been merged for a while now: https://github.com/rust-lang/rust/pull/44374

**Comment by nikhilm** [comment]
Sweet! Thank you!

>

**Pull request closed without merging**